File: python/src/protoss_orchestrated.py
import hashlib
from dataclasses import dataclass

import sodium_bindings as sodium
from protoss_protocol import (
    ReturnTypeRspDer, hash_to_point, concatenate_bytes,
    POINT_LEN, SCALAR_LEN, SESSION_KEY_LEN,
)


@dataclass
class ProtossOrchestratedState:
    secret_scalar: bytes
    I: bytes
    P_i: bytes
    P_j: bytes
    V: bytes


def orchestrated_state_create(P_i: bytes, P_j: bytes) -> ProtossOrchestratedState:
    """Create a state with party identifiers"""
    return ProtossOrchestratedState(
        secret_scalar=bytes(SCALAR_LEN),
        I=bytes(POINT_LEN),
        P_i=P_i,
        P_j=P_j,
        V=bytes(POINT_LEN),
    )


def orchestrated_state_destroy(state: ProtossOrchestratedState) -> None:
    """Overwrite the secret scalar"""
    state.secret_scalar = bytes(SCALAR_LEN)
    state.V = bytes(POINT_LEN)


def orchestrated_Init(state: ProtossOrchestratedState, password: str) -> bytes:
    """Initialize protocol (Step 1): fills state, returns I"""
    state.secret_scalar = sodium.crypto_core_ristretto255_scalar_random()
    X = sodium.crypto_scalarmult_ristretto255_base(state.secret_scalar)
    state.V = hash_to_point(password)
    state.I = sodium.crypto_core_ristretto255_add(X, state.V)
    return state.I


def orchestrated_RspDer(state: ProtossOrchestratedState, password: str, I: bytes) -> ReturnTypeRspDer:
    """Response and key derivation (Step 2): fills state, returns R and K"""
    state.secret_scalar = sodium.crypto_core_ristretto255_scalar_random()
    Y = sodium.crypto_scalarmult_ristretto255_base(state.secret_scalar)
    state.V = hash_to_point(password)
    R = sodium.crypto_core_ristretto255_add(Y, state.V)
    X_prime = sodium.crypto_core_ristretto255_sub(I, state.V)
    Z = sodium.crypto_scalarmult_ristretto255(state.secret_scalar, X_prime)

    concat = concatenate_bytes([Z, I, R, state.P_i, state.P_j, state.V])
    K = hashlib.sha512(concat).digest()[:SESSION_KEY_LEN]

    return ReturnTypeRspDer(R, K)


def orchestrated_Der(state: ProtossOrchestratedState, R: bytes) -> bytes:
    """Key derivation (Step 3)"""
    Y_prime = sodium.crypto_core_ristretto255_sub(R, state.V)
    Z = sodium.crypto_scalarmult_ristretto255(state.secret_scalar, Y_prime)

    concat = concatenate_bytes([Z, state.I, R, state.P_i, state.P_j, state.V])
    K = hashlib.sha512(concat).digest()[:SESSION_KEY_LEN]

    return K

File: dalek-rust/src/protoss_orchestrated.rs
#![forbid(unsafe_code)]

use curve25519_dalek::{
    ristretto::RistrettoPoint,
    scalar::Scalar,
};
use getrandom::getrandom;
use sha2::{Digest, Sha512};

use crate::ec_operations::{hash_to_point, concatenate_vectors, SCALAR_LEN};
use crate::protoss_protocol::{
    Error, ReturnTypeRspDer, SESSION_ID_BYTES, SESSION_KEY_BYTES,
};

// State managed by the orchestrator and filled by the protocol steps
pub struct ProtossOrchestratedState {
    pub secret_scalar: Scalar,
    pub i: RistrettoPoint,
    pub p_i: [u8; SESSION_ID_BYTES],
    pub p_j: [u8; SESSION_ID_BYTES],
    pub v: RistrettoPoint,
}

// Orchestrator: create a state with party identifiers
pub fn orchestrated_state_create(
    p_i: &[u8; SESSION_ID_BYTES],
    p_j: &[u8; SESSION_ID_BYTES],
) -> ProtossOrchestratedState {
    ProtossOrchestratedState {
        secret_scalar: Scalar::ZERO,
        i: RistrettoPoint::default(),
        p_i: *p_i,
        p_j: *p_j,
        v: RistrettoPoint::default(),
    }
}

// Cleanup: overwrite the secret scalar
pub fn orchestrated_state_destroy(state: &mut ProtossOrchestratedState) {
    state.secret_scalar = Scalar::ZERO;
}

// Initialize protocol (Step 1): fills state by reference, returns I
pub fn orchestrated_init(
    state: &mut ProtossOrchestratedState,
    password: &str,
) -> Result<RistrettoPoint, Error> {
    let mut x_bytes = [0u8; SCALAR_LEN];
    getrandom(&mut x_bytes)?;
    state.secret_scalar = Scalar::from_bytes_mod_order(x_bytes);

    let x_point = RistrettoPoint::mul_base(&state.secret_scalar);
    state.v = hash_to_point(password);
    state.i = x_point + state.v;

    Ok(state.i)
}

// Response and key derivation (Step 2): fills state by reference, returns R and K
pub fn orchestrated_rsp_der(
    state: &mut ProtossOrchestratedState,
    password: &str,
    i: RistrettoPoint,
) -> Result<ReturnTypeRspDer, Error> {
    let mut y_bytes = [0u8; SCALAR_LEN];
    getrandom(&mut y_bytes)?;
    state.secret_scalar = Scalar::from_bytes_mod_order(y_bytes);

    let y_point = RistrettoPoint::mul_base(&state.secret_scalar);
    state.v = hash_to_point(password);
    let r = y_point + state.v;
    let x_prime = i - state.v;
    let z = state.secret_scalar * x_prime;

    let concat = concatenate_vectors(&[
        z.compress().as_bytes(),
        i.compress().as_bytes(),
        r.compress().as_bytes(),
        state.p_i.as_ref(),
        state.p_j.as_ref(),
        state.v.compress().as_bytes(),
    ]);
    let mut hasher = Sha512::new();
    hasher.update(&concat);
    let hash = hasher.finalize();
    let mut k = [0u8; SESSION_KEY_BYTES];
    k.copy_from_slice(&hash[..SESSION_KEY_BYTES]);

    Ok(ReturnTypeRspDer { r, k })
}

// Key derivation (Step 3)
pub fn orchestrated_der(
    state: &ProtossOrchestratedState,
    r: RistrettoPoint,
) -> Result<[u8; SESSION_KEY_BYTES], Error> {
    let y_prime = r - state.v;
    let z = state.secret_scalar * y_prime;

    let concat = concatenate_vectors(&[
        z.compress().as_bytes(),
        state.i.compress().as_bytes(),
        r.compress().as_bytes(),
        state.p_i.as_ref(),
        state.p_j.as_ref(),
        state.v.compress().as_bytes(),
    ]);
    let mut hasher = Sha512::new();
    hasher.update(&concat);
    let hash = hasher.finalize();
    let mut k = [0u8; SESSION_KEY_BYTES];
    k.copy_from_slice(&hash[..SESSION_KEY_BYTES]);

    Ok(k)
}
